src/graphics/device-cache.js

```diff
@@ -0,0 +1,39 @@
+/** @typedef {import('./graphics-device.js').GraphicsDevice} GraphicsDevice */
+
+/**
+ * A cache storing shared resources associated with a device. The resources are removed
+ * from the cache when the device is destroyed.
+ *
+ * @ignore
+ */
+class DeviceCache {
+    /**
+     * Cache storing the resource for each GraphicsDevice
+     *
+     * @type {Map<GraphicsDevice, any}
+     */
+    _cache = new Map();
+
+    /**
+     * Returns the resources for the supplied device.
+     *
+     * @param {GraphicsDevice} device - The graphics device.
+     * @returns {any} The resource for the device.
+     */
+    get(device, onCreate) {
+
+        if (!this._cache.has(device)) {
+            this._cache.set(device, onCreate());
+
+            // when the device is destroyed, destroy and remove its entry
+            device.on('destroy', () => {
+                this._cache.get(device)?.destroy();
+                this._cache.delete(device);
+            });
```

Comment:
Agree it's not ideal to access _cache twice, but it's unnecessary:

```suggestion
            const result = onCreate();
            this._cache.set(device, result);

            // when the device is destroyed, destroy and remove its entry
            device.on('destroy', () => {
                this._cache.get(device)?.destroy();
                this._cache.delete(device);
            });
            return result;
```

Reply:
I'm not talking about this once off case .. I'm taking about the normal case where cache contains the value already.
first we need to do 'cache.has'
and then do 'cache.get'

before we only did cache.get and compared to null .. (which limited storage of null)

Reply:
yes true. if perf were an issue you could rather do:
```
const value = this._cache.get(device);
if (!value && !this._cache.has(device)) {
    this._cache.set(blah);
    etc
}
return value;